pkg/addons/highavailability.go

```diff
@@ -2,99 +2,105 @@ package addons
-	loading := spinner.Start()
-	defer loading.Close()
-
-	logrus.Debugf("Enabling high availability")
+func EnableHA(
+	ctx context.Context, kcli client.Client, kclient kubernetes.Interface, hcli helm.Client,
+	isAirgap bool, serviceCIDR string, proxy *ecv1beta1.ProxySpec, cfgspec *ecv1beta1.ConfigSpec,
+	logMsg func(string, ...any),
+) error {
+	// Log to both the user and the logrus logger so that the logs appear in the log file.
+	logFn := func(format string, args ...any) {
+		logMsg(format, args...)
+		logrus.Debugf(format, args...)
+	}
 
 	if isAirgap {
-		loading.Infof("Enabling high availability")
+		logFn("Enabling high availability")
 
-		domains := runtimeconfig.GetDomains(cfgspec)
-
-		// TODO (@salah): add support for end user overrides
-		sw := &seaweedfs.SeaweedFS{
-			ServiceCIDR:         serviceCIDR,
-			ProxyRegistryDomain: domains.ProxyRegistryDomain,
-		}
-		exists, err := hcli.ReleaseExists(ctx, sw.Namespace(), sw.ReleaseName())
+		hasMigrated, err := registry.IsRegistryHA(ctx, kcli)
 		if err != nil {
-			return errors.Wrap(err, "check if seaweedfs release exists")
-		}
-		if !exists {
+			return errors.Wrap(err, "check if registry data has been migrated")
+		} else if !hasMigrated {
 			logrus.Debugf("Installing seaweedfs")
-			if err := sw.Install(ctx, kcli, hcli, addOnOverrides(sw, cfgspec, nil), nil); err != nil {
-				return errors.Wrap(err, "install seaweedfs")
+			err = ensureSeaweedfs(ctx, kcli, hcli, serviceCIDR, cfgspec)
+			if err != nil {
+				return errors.Wrap(err, "ensure seaweedfs")
 			}
 			logrus.Debugf("Seaweedfs installed!")
-		} else {
-			logrus.Debugf("Seaweedfs already installed")
-		}
 
-		// TODO (@salah): add support for end user overrides
-		reg := &registry.Registry{
-			ServiceCIDR:         serviceCIDR,
-			ProxyRegistryDomain: domains.ProxyRegistryDomain,
-			IsHA:                true,
-		}
-		logrus.Debugf("Migrating registry data")
-		if err := reg.Migrate(ctx, kcli, loading); err != nil {
-			return errors.Wrap(err, "migrate registry data")
-		}
-		logrus.Debugf("Registry migration complete!")
-		logrus.Debugf("Upgrading registry")
-		if err := reg.Upgrade(ctx, kcli, hcli, addOnOverrides(reg, cfgspec, nil)); err != nil {
-			return errors.Wrap(err, "upgrade registry")
+			logrus.Debugf("Scaling registry to 0 replicas")
+			// if the migration fails, we need to scale the registry back to 1
+			defer maybeScaleRegistryBackOnFailure(kcli)
+			err := scaleRegistryDown(ctx, kcli)
+			if err != nil {
+				return errors.Wrap(err, "scale registry to 0 replicas")
+			}
+
+			logFn("Migrating data for high availability")
+			err = migrateRegistryData(ctx, kcli, kclient, cfgspec, logFn)
+			if err != nil {
+				return errors.Wrap(err, "migrate registry data")
+			}
+			logrus.Debugf("Data migration complete!")
+
+			logFn("Enabling registry high availability")
```

Comment:
@ajp-io can you review this message to the user? thanks

Reply:
```suggestion
			logFn("Enabling high availability for the registry")
```